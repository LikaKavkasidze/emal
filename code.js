/// Emal – Acquire and process scientific data with ease
/// code.js – Emal code lexer and interpreter.
///
/// Copyright (C) 2023 Titouan (Stalone) S.
/// This program is free software: you can redistribute it and/or modify it
/// under the terms of the GNU Affero General Public License as published by
/// the Free Software Foundation, version 3.
///
/// The code lexer is a very simple state machine running on each line of
/// the user code. Each line expects a strict order: keyword, variable,
/// potential unit, equal sign and value or expression. Invalid lines are
/// discarded.

import { StateMachine } from "./stateMachine.js";
import { Expression } from "./expression.js";
import { isWhitespace } from "./matchers.js";

const VARIABLE_KW = "let";

function start(charCode) {
	// Eat leading spaces
	if(isWhitespace(charCode)) return start;

	this.token({}, "keyword");
	return this.keyword(charCode);
}

function keyword(charCode) {
	if(isWhitespace(charCode)) return this.afterKeyword(charCode);

	this.eat(charCode);
	return keyword;
}

function afterKeyword(charCode) {
	if(isWhitespace(charCode)) return afterKeyword;

	this.token({}, "variable");
	return this.variable(charCode);
}

function variable(charCode) {
	// 61 = "="
	if(charCode === 61) return this.equalSign(charCode); 
	if(isWhitespace(charCode)) return this.beforeEqual(charCode);
	// 95 = "_"
	if(charCode === 95) {
		this.token({}, "unit");
		return unit;
	}

	this.eat(charCode);
	return variable;
}

function unit(charCode) {
	// 61 = "="
	if(charCode === 61) return this.equalSign(charCode); 
	if(isWhitespace(charCode)) return this.beforeEqual(charCode);

	this.eat(charCode);
	return unit;
}

function beforeEqual(charCode) {
	if(charCode === 61) return this.equalSign(charCode);

	return beforeEqual;
}

function equalSign(charCode) {
	if(charCode !== 61) return this.nok(charCode);

	return afterEqual;
}

function afterEqual(charCode) {
	if(isWhitespace(charCode)) return afterEqual;

	this.token({}, "expression");
	return this.expression(charCode);
}

function expression(charCode) {
	if(isNaN(charCode)) return this.end(charCode);

	this.eat(charCode);
	return expression;
}

function toValue(token) {
	return token.value;
}

const lineSm = new StateMachine("start", {
	start,
	keyword,
	afterKeyword,
	variable,
	unit,
	beforeEqual,
	equalSign,
	afterEqual,
	expression,
}, {
	type: "parallel",
	transformers: {
		keyword: toValue,
		variable: toValue,
		unit: toValue,
		expression: tok => new Expression(tok.value),
	},
});

export class CodeBlock {
	constructor(codeStr) {
		this.lines = codeStr.split("\n").reduce((acc, line) => {
			const runnerOutput = lineSm.run(line).transform();
			if(!runnerOutput.hasThrown()) acc.push(runnerOutput.tokens);

			return acc;
		}, []);
	}

	// For now, evaluation simply consists of evaluating the expression
	// for each variable.
	evaluate(userData) {
		const generatedData = {};

		this.lines.forEach(line => {
			switch(line.keyword) {
				case VARIABLE_KW:
					const data = Object.assign({}, userData, generatedData);
					generatedData[line.variable] = line.expression.evaluate(data);
					break;
			}
		});

		return generatedData;
	}
}
