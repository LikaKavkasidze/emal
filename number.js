/// Emal – Acquire and process scientific data with ease
/// number.js – Parser of arbitrary precision numbers.
///
/// An Emal number is stored as a large integer (BigInt) with decimal
/// place position.
/// Parsing is based on a state machine with states corresponding to
/// parts of the number.

import { StateMachine } from "./stateMachine.js";
import { isDigit, isSign, isComma } from "./matchers.js";

// Default decimal count for `toString`
const ENUMBER_DEFAULT_DC = 2;
// Default extra decimals computed for division
const ENUMBER_DEFAULT_DIV_OFFSET = BigInt(ENUMBER_DEFAULT_DC + 2);

// Define a start state which defines a local variable and simply returns.
// This declared local variable is used to construct tokens generically, so 
// that parsing of the number and exponent uses the same states.
function start(charCode) {
	this.isExponent = false;
	return this.sign(charCode);
}

function sign(charCode) {
	if(!isSign(charCode)) {
		return this.startIntPart(charCode);
	}

	this.token({}, `${this.isExponent ? "eS" : "s"}ign`);
	this.eat(charCode);

	return startIntPart;
}

function startIntPart(charCode) {
	if(!isDigit(charCode)) {
		return this.nok(charCode);
	}

	this.token({}, `${this.isExponent ? "eI" : "i"}ntPart`);
	this.eat(charCode);

	return intPart;
}

function intPart(charCode) {
	if(charCode === 69 || charCode === 101) {
		this.isExponent = true;
		return sign;
	// Exponents cannot have a floating part
	} else if(isComma(charCode) && !this.isExponent) {
		this.token({}, "floatPart");
		return floatPart;
	} else if(isDigit(charCode)) {
		this.eat(charCode);
		return intPart;
	}

	return this.end(charCode);
}

function floatPart(charCode) {
	if(charCode === 69 || charCode === 101) {
		this.isExponent = true;
		return sign;
	} else if(isDigit(charCode)) {
		this.eat(charCode);
		return floatPart;
	}

	return this.end(charCode);
}

// Used as a transformer on token values which are strings
function toNumber(token) {
	return BigInt(token.value);
}

function toValue(token) {
	return token.value;
}

export const numberSm = new StateMachine("start",
	{
		start,
		sign,
		startIntPart,
		intPart,
		floatPart,
	},
	{
		transformers: {
			sign: toValue,
			intPart: toValue,
			floatPart: toValue,
			eSign: toValue,
			eIntPart: toNumber,
		},
		type: "parallel",
	},
);

function BIPow(n, p) {
	let o = n;

	for(; p > 1; p--) {
		o *= n;
	}

	return o;
}

function preprocessOp(a, b) {
	let aInt = a.rawInt;
	let bInt = b.rawInt;

	let aLen = a.decimalPlace;
	let bLen = b.decimalPlace;

	if(aLen > bLen) {
		const delta = aLen - bLen;
		bInt *= BIPow(10n, delta);
		bLen += delta;
	} else if(bLen > aLen) {
		const delta = bLen - aLen;
		aInt *= BIPow(10n, delta);
		aLen += delta;
	}

	return {
		aInt,
		bInt,
		oLen: aLen,
	};
}

export class EmalNumber {
	constructor(rawInt, decimalPlace) {
		this.rawInt = rawInt;
		this.decimalPlace = decimalPlace;
	}

	static fromTokens(userTokens) {
		// Use defaults because some parts of the number are optional
		const defaultTokens = {
			sign: "+",
			floatPart: "0",
			eSign: "+",
			eIntPart: "0",
		};

		const tokens = Object.assign({}, defaultTokens, userTokens);

		const rawInt = BigInt(`${tokens.sign}${tokens.intPart}${tokens.floatPart}`);
		const exponent = BigInt(`${tokens.eSign}${tokens.eIntPart}`);
		const decimalPlace = BigInt(tokens.floatPart.length) - exponent;

		return new EmalNumber(rawInt, decimalPlace);
	}

	static fromString(numberStr) {
		const runnerOutput = numberSm.run(numberStr).transform();
		return EmalNumber.fromTokens(runnerOutput.tokens);
	}

	toString(decimalCount) {
		// Dynamic default argument
		if(typeof decimalCount !== "number") decimalCount = ENUMBER_DEFAULT_DC;

		const { rawInt, decimalPlace } = this;

		const intStr = rawInt.toString(10);
		const exponent = BigInt(intStr.length) - decimalPlace - 1n;

		const intPart = intStr.substr(0, 1);
		const floatPart = intStr.substr(1, decimalCount);

		return `${intPart},${floatPart}${exponent !== 0n ? "e".concat(exponent) : ""}`
	}

	static add(a, b) {
		const { aInt, bInt, oLen } = preprocessOp(a, b);
		return new EmalNumber(aInt + bInt, oLen);
	}

	static sub(a, b) {
		const { aInt, bInt, oLen } = preprocessOp(a, b);
		return new EmalNumber(aInt - bInt, oLen);
	}

	static mul(a, b) {
		const { aInt, bInt, oLen } = preprocessOp(a, b);
		return new EmalNumber(aInt * bInt, oLen * 2n);
	}

	static div(a, b) {
		const { aInt, bInt, oLen } = preprocessOp(a, b);
		// Dividing arbitrary precision numbers cannot be done
		// with infinite precision, here the number is cut to
		// ENUMBER_DEFAULT_DIV_OFFSET digits after decimal point.
		const offset = ENUMBER_DEFAULT_DIV_OFFSET + oLen;
		return new EmalNumber(aInt * BIPow(10n, offset) / bInt, offset);
	}
}
