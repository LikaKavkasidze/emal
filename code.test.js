import { CodeBlock } from "./code.js";
import { EmalNumber } from "./number.js";

test("parses code", () => {
	const cb = new CodeBlock(`let a0 = 2

	let y = x / a0`);

	expect(cb.lines).toMatchSnapshot();
});

test("evaluates code", () => {
	const cb = new CodeBlock(`let a0 = 2

	let y = x / a0`);

	const data = { x: EmalNumber.fromString("1.2") };
	expect(cb.evaluate(data).y.toString()).toEqual("6,00e-1");
});
