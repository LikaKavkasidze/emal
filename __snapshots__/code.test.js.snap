// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses code 1`] = `
[
  {
    "expression": Expression {
      "tokens": [
        {
          "children": EmalNumber {
            "decimalPlace": 1n,
            "isNegative": false,
            "rawInt": 20n,
          },
          "type": "number",
          "value": "",
        },
      ],
    },
    "keyword": "let",
    "variable": "a0",
  },
  {
    "expression": Expression {
      "tokens": [
        {
          "type": "variable",
          "value": "x",
        },
        {
          "type": "variable",
          "value": "a0",
        },
        {
          "type": "operator",
          "value": "/",
        },
      ],
    },
    "keyword": "let",
    "variable": "y",
  },
]
`;
